src/util/retry.rs

```diff
@@ -51,3 +51,30 @@ pub trait RetryPolicy {
 	/// execute the operation that is being retried.
 	fn max_total_delay(&self) -> Option<Duration>;
 }
+
+/// The exponential backoff strategy is a retry approach that doubles the delay between retries.
+pub struct ExponentialBackoff {
+	/// The base delay duration for the backoff algorithm. First retry is `base_delay` after first attempt.
+	base_delay: Duration,
+	/// The maximum number of attempts to retry.
+	max_attempts: u32,
+	/// The maximum accumulated delay that will be allowed over all attempts.
+	/// No upper bound is placed on total delay if `max_total_delay` is None.
+	max_total_delay: Option<Duration>,
+}
+
+impl RetryPolicy for ExponentialBackoff {
+	fn next_delay(&self, attempts_made: u32) -> Duration {
+		let backoff_factor = 2_u32.pow(attempts_made.min(self.max_attempts)) - 1;
+		let delay = self.base_delay * backoff_factor;
+		delay
+	}
+
+	fn max_attempts(&self) -> u32 {
+		self.max_attempts
+	}
+
+	fn max_total_delay(&self) -> Option<Duration> {
+		self.max_total_delay
+	}
+}
```

Comment:
The idea is pretty much the same as `Iteratror`, FWIW. The current interface seems to presume aspects of a policy in the interface, which may not be applicable for all implementations. Thus, you have `Option` creeping into `max_total_delay` and no way of not having a maximum number other than returning an unreasonably large number.

Reply:
> Doesn't similar thing hold for `accumulated_delay` leaking/creeping in next_delay when someone doesn't want to have `max_total_delay`?

Hmm... I don't see it that way. It's more the current context for the `RetryPolicy` to make a decision. The `retry` method would be responsible for making attempts, keeping track of how many were made and accumulated delay. With this, it would simply ask the `RetryPolicy`, "What now?" They each have their own responsibilities whereas the current approach has the policy logic leaking back into the retrier.

> But there is no default impl currently.
> 

Nor is there one for `Iterator`, but it has default (i.e., provided) methods which consume the iterator, wrapping it with a decorator that also implements `Iterator`. That is how you are able to chain iterators like: `v.iter().filter().map()`. Each method is returning a type implementing `Iterator`, which has it's own responsibility.

So you could have something like:

```rust
ExponentialBackoffRetryPolicy::new(Duration::from_secs(2))
	.with_max_attempts(5)
	.with_max_total_delay(Duration::from_secs(60))
	.with_max_jitter(Duration::from_micros(100));
```

Each call is simply decorating the previously returned `RetryPolicy`, delegating to the wrapped one when appropriate. The default methods are providing ways to decorate other implementations with similar behaviors.

Note: Those methods are defined on `RetryPolicy`, not directly on `ExponentialBackoffRetryPolicy`.

> RetryPolicy in itself doesn't have max_attempts and max_total_delay as members.

It wouldn't need them. You'd have `MaxAttemptsRetryPolicy` and `MaxTotalDelayRetryPolicy` structs that would. Then if someone comes around and adds `SomeNewFancyRetryPolicy`, they get max attempts and total delay for free, if they want it.

> 
> it is `ExponentialBackoff` impl detail to have it as const members. `JitteredRetryPolicy` for example returns innerPolicy's values.

That's essentially how the approach I'm suggesting would work, only more generally. So `MaxAttemptsRetryPolicy` would return `None` if the given attempts exceeds the max. Otherwise, it would delegate to the wrapped one, which in turn may do something similar. `JitteredRetryPolicy` would unconditionally call it's wrapped version and add the jitter before returning.

This way, each policy can be encapsulated in a small struct implementing one method for _one_ specific behavior.

Reply:
> Few problems with above approach.
> 
> * accumulated_delay is sent to policies which don't need it, but it is ok.

Yeah, can't really avoid this. I just think of it as the retrier's context. Could even include the error and wrap everything in a struct to make it so it's not directly exposed in the interface.

> * i want to keep `MaxAttemptsRetryPolicy` `MaxTotalDelayRetryPolicy` and `JitteredRetryPolicy` pub(crate) or private and allow users to use with_* functions, is that doable ? (otherwise they just create more confusion)
>   Afaiu, i have to keep returning concrete types, and it is not possible to return generic trait from with_* functions.

Correct, they need to be `pub` similar how the standard library has a `Filter` struct even though it is never instantiated directly by the user.

https://doc.rust-lang.org/std/iter/struct.Filter.html

Should be fine and you can just make them not constructable from outside the crate.

Reply:
> Ahh, i see. So what you are suggesting is i move from kind-of decorator pattern to decorator-builder pattern. Afaiu, it doesn't work well with bindings, and that is the reason we didn't do this in LDK. (If it doesn't work well with bindings it might not be a good option)

Could just make a separate interface for bindings that wraps these. Maybe something with `Arc`s? Not ideal, but the decorator pattern is really nice in rust.

Reply:
Yeah, that looks right.